os/drivers/testcase/test_sem.c

```diff
@@ -0,0 +1,136 @@
+/****************************************************************************
+ *
+ * Copyright 2019 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <debug.h>
+#include <time.h>
+#include <semaphore.h>
+#include <tinyara/sched.h>
+#include <clock/clock.h>
+#include <tinyara/testcase_drv.h>
+#include "kernel_test_proto.h"
+
+/****************************************************************************
+ * Private Function
+ ****************************************************************************/
+
+int test_sem_tick_wait(int cmd, unsigned long arg)
+{
+	int ret_chk;
+	sem_t sem;
+	struct timespec cur_time;
+	struct timespec base_time;
+
+	/* init sem count to 1 */
+
+	ret_chk = sem_init(&sem, 0, 1);
+	if (ret_chk != OK) {
+		dbg("sem_init failed.");
+		return ERROR;
+	}
+
+	/* success to get sem case test */
+
+	ret_chk = clock_gettime(CLOCK_REALTIME, &base_time);
+	if (ret_chk != OK) {
+		dbg("clock_gettime failed.");
+		return ERROR;
+	}
+
+	ret_chk = sem_tickwait(&sem, clock(), 2);
+	if (ret_chk != OK) {
+		dbg("sem_tickwait failed.");
+		return ERROR;
+	}
+
+	ret_chk = clock_gettime(CLOCK_REALTIME, &cur_time);
+	if (ret_chk != OK) {
+		dbg("clock_gettime failed.");
+		return ERROR;
+	}
+	if (base_time.tv_sec + 2 == cur_time.tv_sec) {
+		dbg("sem_timedwait failed.");
+		return ERROR;
+	}
+
+	ret_chk = sem_post(&sem);
+	if (ret_chk != OK) {
+		dbg("sem_post failed.");
+		return ERROR;
+	}
+
+	ret_chk = sem_destroy(&sem);
+	if (ret_chk != OK) {
+		dbg("sem_destroy failed.");
+		return ERROR;
+	}
+
+	/* init sem count to 0 */
+
+	ret_chk = sem_init(&sem, 0, 0);
+	if (ret_chk != OK) {
+		dbg("sem_init failed.");
+		return ERROR;
+	}
+
+	/* expired time test */
+
+	ret_chk = sem_tickwait(&sem, clock() - 2, 0);
+	if (ret_chk != ERROR) {
+		dbg("sem_tickwait failed.");
+		return ERROR;
```

Comment:
I think we'd better add sem_destory for error case.

Reply:
I agree.
I will add sem_destroy for error case